src/lib/registration/controllers/package_search.rb

```diff
@@ -0,0 +1,180 @@
+# Copyright (c) [2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "registration/package_search"
+
+Yast.import "Popup"
+Yast.import "HTML"
+
+module Registration
+  module Controllers
+    # Implements the actions and keeps the state for the package search feature
+    class PackageSearch
+      include Yast::I18n
+
+      # @return [Array<RemotePackage>] List of selected packages
+      attr_reader :selected_packages
+
+      # Constructor
+      def initialize
+        textdomain "registration"
+        @selected_packages = []
+      end
+
+      # Performs a package search
+      #
+      # @param text [String] Term to search for
+      # @return [Array<Registration::RemotePackage>] List of packages
+      def search(text)
+        @search = ::Registration::PackageSearch.new(text: text)
+        @search.packages
+      end
+
+      # Selects/unselects the current package for installation
+      #
+      # It does nothing if the package is already installed.
+      def toggle_package(package)
+        return if package.installed?
+
+        if package.selected?
+          unselect_package(package)
+        else
+          select_package(package)
+        end
+      end
+
+    private
+
+      # Selects the current package for installation
+      #
+      # If required, it selects the corresponding addon
+      #
+      # @param package [RemotePackage] Package to select
+      def select_package(package)
+        addon = package.addon
+        select_addon(addon) if addon
+        set_package_as_selected(package) if addon.nil? || addon.selected? || addon.registered?
```

Comment:
NP: this condition looks a bit duplicate and complex. What do you think about having select_addon that return boolean ( now only partially) that looks something like:
```
def select_addon(addon)
    return true if addon.nil? || addon.registered? || addon.selected? || addon.auto_selected?
￼    addon.selected if enable_addon?(addon)
￼end
```

Also that brings me to another question. What happen here if addon is just auto_selected? it looks like it skips setting package as selected, not?

Reply:
Actually, that was a bug. If an addon is marked as '#auto_selected', it should be marked as '#selected' so it does not get removed accidentally. Fixed now.